File: STROPR.py
def modpod(base,exp,mod):
	if exp==1:
		return base
	if exp&1:
		return (base*(modpod(base,(exp-1)/2,mod)**2))%mod

	return ((modpod(base,exp/2,mod)**2))%mod


def main():
	mod = (10**9)+7
	calc = []
	for i in xrange((10**5)+1):
		calc.append(modpod(i,mod-2,mod))

	for i in xrange(input()):
		n,x,m = map(int,raw_input().split())
		lis = map(int,raw_input().split())

		sumi = 0

		pre = 1
		for i in xrange(x):
			sumi = (sumi+(pre*lis[x-1-i]))%mod
			pre = (pre*(m+i))%mod
			pre = (pre*calc[i+1])%mod
 
		print sumi%mod
main()

File: STROPR.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long int ll;


ll mod = 1000000007;
ll power(ll a,ll b)
{
	if (b==1)
		return a;

	if(b&1)
	{
		ll base = (long long int)(a*(pow(power(a,(b-1)/2),2)))%mod;
		return base%mod;
	}
	else
	{
		ll base = (long long int)(pow(power(a,b/2),2))%mod;
		return base%mod;
	}
}

main()
{
	freopen("input.txt","r",stdin);
	freopen("output.txt","w",stdout);
	cout<<"hehe"<<" "<<power(2,mod-2)<<"\n";
	ll t;
	scanf("%lld",&t);

	while(t--)
	{
		ll n,x,m;
		scanf("%lld%lld%lld",&n,&x,&m);

		ll arr[n+5];
		for(ll i=0;i<n;i++)
			scanf("%lld",&arr[i]);

		ll pre = 1,sum=0;
		for(ll i=0;i<x;i++)
		{
			cout<<pre<<" "<<power(i+1,mod-2)<<"\n";
			sum = (sum+(pre*arr[x-i-1]))%mod;
			pre = (pre*(m+i))%mod;
			pre = (pre*power(i+1,mod-2))%mod;
		}

		printf("%lld\n",sum%mod );
	} 

}